### ISSUE TITLE
Canonical tag for root page has form http://www.example.com/1

### ISSUE THREAD

**Issue by thomaszbz** [mblaschke/TYPO3-metaseo#224]
Root page produces canonical tag

```
<link rel="canonical" href="http://www.example.com/1">
```

instead of

```
<link rel="canonical" href="http://www.example.com">
```

Related: #203

**Comment by thomaszbz** [comment]
With the patch applied, I still get `/1`.

With a root page's `id=2`, this happens with empty canonical url in page properties:
```
select tx_metaseo_canonicalurl, deleted, hidden from pages where uid=2;
+-------------------------+---------+--------+
| tx_metaseo_canonicalurl | deleted | hidden |
+-------------------------+---------+--------+
|                         |       0 |      0 |
+-------------------------+---------+--------+
```

![auswahl_066](https://cloud.githubusercontent.com/assets/11200727/20870841/085724cc-ba8f-11e6-9c99-3ac3fcdc5e7d.png)


Given that it's empty, and the patch considers only cases where it's [not empty](https://github.com/mblaschke/TYPO3-metaseo/pull/267/commits/43c28e8c37218db52c4105b9e98783a38b33c055#diff-fd0253a2fd2951af5565190c2b11ed51R1334), the patch at least does not solve the issue I have. Maybe it solves other cases. Eventually it solves #203 ? @dsteinborn could you please report back which case the PR #267 is supposed to solve?

**Comment by thomaszbz** [comment]
This bug still exists in 6.2.29, Realurl 2.0.15 and Realurl 2.1.5.

Another page with pretty much the same setup (including Typo3 version and Realurl version) does not have this issue.

**Comment by thomaszbz** [comment]
I further tracked it down: First, I found multiple entries for my root page:
![auswahl_067](https://cloud.githubusercontent.com/assets/11200727/20871336/13ae92be-ba95-11e6-8f5b-d13e18069ab0.png)

Then I deleted these entries, logged out from the backend and http-requested the root page for indexation. As a result, I got a new entry. Note that the new entry somehow got the URL `1` which is completely wrong (even the uid would be 2).
![auswahl_068](https://cloud.githubusercontent.com/assets/11200727/20871364/4d27bed0-ba95-11e6-8f2a-81ab9e5d465c.png)

```
select uid, page_uid, page_url from tx_metaseo_sitemap where page_uid=2;
+-----+----------+----------+
| uid | page_uid | page_url |
+-----+----------+----------+
| 897 |        2 | 1        |
+-----+----------+----------+


```

Basically that means that this issue happens while the index entry is generated, not when meta tags are generated. Which explains why I have this issue on one instance and not on another.

**Comment by thomaszbz** [comment]
On the affected TYPO3 instance, I added some debug output to `metaseo/Classes/Hook/SitemapIndexLinkHook.php `:


    public function hook_linkParse(array &$pObj)
    {
        if (!$this->checkIfSitemapIndexingIsEnabled('typolink')) {
            return;
        }

        // Check
        if (empty($pObj['finalTagParts']) || empty($pObj['conf']) || empty($pObj['finalTagParts']['url'])) {
            // no valid link
            return;
        }

        var_dump($pObj['finalTagParts']['url']);

I tried again, and for URL `https://www.example.com` I got the output:

```
string(18) "ueber-uns/kontakt/"
string(18) "ueber-uns/kontakt/"
string(1) "1"
string(23) "https://www.example.com/1"
string(23) "https://www.example.com/1"
string(1) "1"
string(1) "1" 
```

Basically, that means that this `1` stuff comes directly from TYPO3 6.2.29 or Realurl 2.1.5. It's not a bug in Metaseo in the first place.

**Comment by thomaszbz** [comment]
In TYPO3 6.2, injection of $pObj looks like this:

			// Hook: Call post processing function for link rendering:
			if (isset($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_content.php']['typoLink_PostProc']) && is_array($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_content.php']['typoLink_PostProc'])) {
				$_params = array(
					'conf' => &$conf,
					'linktxt' => &$linktxt,
					'finalTag' => &$res,
					'finalTagParts' => &$finalTagParts
				);
				foreach ($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_content.php']['typoLink_PostProc'] as $_funcRef) {
					GeneralUtility::callUserFunction($_funcRef, $_params, $this);
				}
			}

**Comment by thomaszbz** [comment]
It comes from RealURL (table `tx_realurl_urldata`):

```
select request_variables, speaking_url from tx_realurl_urldata where speaking_url=1;
+--------------------+---------------------------+
| request_variables  | speaking_url              |
+--------------------+---------------------------+
| {"id":"2","L":"0"} | 1                         |
| {"id":"2"}         | 1                         |
| {"L":"1","id":2}   | 1/                        |
| {"L":"1","id":28}  | 1/ueber-uns/firmenprofil/ |
| {"L":"1","id":53}  | 1/ueber-uns/impressum/    |
+--------------------+---------------------------+
5 rows in set (0.00 sec)

```

Eventually these are old entries from older versions of RealUrl.

**Comment by thomaszbz** [comment]
I checked for duplicate title, duplicate url paths, etc. There's no reason for the /1.

I ended up to just change it in the database:

    update tx_realurl_urldata set speaking_url='/' where speaking_url='1' ;

That fixes it: RealUrl does not add new or wrong entries any more, and Metaseo's sitemap is correctly filled with '/' for the root page.

After that, the canonical URL reads:

    <link rel="canonical" href="https://www.example.com/">

Great.

Closing this issue, as it's not a bug in metaseo in any case. And the problem can be worked around.
